---
title: "Enrichment Analysis"
subtitle: "211130_Q96K97del_A603fs dataset"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    theme: flatly
    toc: true
    toc_float: true
    toc_depth: 4
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE, 
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Setup 

```{r packages}
suppressPackageStartupMessages({
  ## Common
  library(tidyverse)
  library(magrittr)
  library(future.apply)
  library(here)
  library(AnnotationHub)
  library(purrr)
  library(scales)
  library(kableExtra)
  library(tictoc)
  library(ggrepel)
  library(RColorBrewer)
  library(ggpubr)
  library(pander)
  library(rmarkdown)
  ## Project specific
  
})
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores <- availableCores() - 1
```

```{r funcs}
source("~/bioinformatics/bioToolkit/lbFuncs.R")
```

```{r projDir}
## Choose either local or remote project directory
# projDir <- here()
projDir <- "/hpcfs/users/a1647910/211130_Q96K97del_A603fs"
```

## EnsDb

```{r drChrs}
drChrs <- seq(1:25)
```

```{r ah}
ah <- AnnotationHub() %>%
  subset(species == "Danio rerio") %>%
  subset(rdataclass == "EnsDb")
ensDb <- ah[["AH83189"]] ## Ens101
genes <- genes(ensDb, filter = SeqNameFilter(drChrs))
mcols(genes) <- mcols(genes)[
  c("gene_id", "gene_name", "gene_biotype", "entrezid")
]
exons <- exonsBy(ensDb, by = "gene", filter = SeqNameFilter(drChrs))
```

An `EnsDb` object for Ensembl release 101 was accessed and gene and exon annotation information was extracted.

## Metadata

```{r meta}
metadata <- read_csv(here("files/samples.csv")) %>%
  ## We need some sample aliases that follow R naming conventions
  mutate(
    alias = case_when(
      genotype == "WT" ~ paste0("WT_", sample),
      genotype == "EOfAD-like" ~ paste0("fAD_", sample),
      genotype == "MPS-IIIB" ~ paste0("MPS_", sample),
    ),
    genotype = factor(genotype, levels = c("WT", "EOfAD-like", "MPS-IIIB")),
    group = genotype
  ) %>%
  dplyr::arrange(genotype) %>%
  dplyr::select(sample, genotype, alias, everything())
genoCols <- metadata$genotype %>%
  levels() %>%
  length() %>%
  brewer.pal("Set1") %>%
  setNames(levels(metadata$genotype))
compCols <- genoCols[2:3]
```

```{r samples_byGeno}
samples_byGeno <- metadata %>%
  split(f = .$genotype) %>%
  sapply(function(x){
    pull(x, sample)
  }, simplify = FALSE)
```

# DE data

```{r deData}
topDE <- readRDS(here("files/topTables.Rds"))
```

```{r deGenes}
deGenes <- sapply(topDE, dplyr::filter, DE, simplify = FALSE)
```