---
title: "Further Enrichment Analyses"
subtitle: "211130_Q96K97del_A603fs dataset"
author: "Lachlan Baer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
    fig_width: 8
    fig_height: 6
    fig_align: "center"
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  error = FALSE,
  fig.align = "center"
)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Reproducibility of this analysis

Please note that newer versions of `babelgene` and potentially `msigdbr` cause issues with the reproducibility of this document.
The issues lies within the `babelgene` package (a dependency of `msigdbr`), which pulls the mappings of zebrafish orthologs to human genes, as required for the construction of gene sets.
Unfortunately the `msigdbr()` function does not possess arguments to control this behaviour in a reproducible manner, and therefore the package version dictates the output of the function.
To successfully reproduce this analysis the user's system must have `babelgene v22.3` installed, and also potentially `misgdbr v7.4.1` (analysis successfully reproduced with this version but not tested with newer versions i.e. 7.5.1).
This document was rendered using a docker container with the required package versions installed (see Session information below for a complete list of packages).
Packages were installed using the `R/docker-setup.R` script on a fresh container spun up from the `bioconductor/bioconductor_docker:RELEASE_3_16` image.

# Setup

```{r packages}
suppressPackageStartupMessages({
  ## Common
  library(tidyverse)
  library(magrittr)
  library(future.apply)
  library(here)
  library(AnnotationHub)
  library(purrr)
  library(scales)
  library(kableExtra)
  library(tictoc)
  library(ggrepel)
  library(RColorBrewer)
  library(ggpubr)
  library(pander)
  library(rmarkdown)
  ## Project specific
  library(edgeR)
  library(limma)
  library(pheatmap)
  library(cqn)
  library(DT)
  library(htmltools)
  library(msigdbr)
  library(fgsea)
  library(plyranges)
})
```

```{r options}
if (interactive()) setwd(here::here())
theme_set(theme_bw())
cores <- availableCores() - 1
```

```{r}
source(here("files/lbFuncs.R"))
```

## EnsDb

```{r ensParams}
ens_species <- "Danio rerio"
ens_release <- "101"
ens_assembly <- "GRCz11"
```

```{r ensDb}
ah <- AnnotationHub() %>%
  subset(species == ens_species) %>%
  subset(rdataclass == "EnsDb")
ahId <- ah$ah_id[str_detect(ah$title, ens_release)]
ensDb <- ah[[ahId]]
```

```{r chrInfo}
chrInfo <- getChromInfoFromEnsembl(ens_assembly, release = ens_release) %>%
  dplyr::filter(coord_system == "chromosome")
primary_chrs <- chrInfo$name
```

```{r genes}
genes <- genes(ensDb, filter = SeqNameFilter(primary_chrs)) %>%
  .[,c("gene_id", "gene_name", "gene_biotype", "entrezid")]
```

An `EnsDb` object was obtained for Ensembl release 101 with the `AnnotationHub` package.
This contained the information required to set up gene and transcript annotations.
Gene-level estimates of GC content and length were also generated which may be required as covariates in suitable analyses.

```{r entrezGenes}
entrezGenes <- genes %>%
  as.data.frame() %>%
  dplyr::filter(!is.na(entrezid)) %>%
  mutate(entrezid = unAsIs(entrezid)) %>%
  unnest(entrezid) %>%
  dplyr::rename(entrez_gene = entrezid)
```

```{r geneChrs}
geneChrs <- as_tibble(genes) %>%
  dplyr::select(gene_id, chromosome = seqnames)
```

## Metadata

```{r meta}
metadata <- read_csv(here("files/samples.csv")) %>%
  ## We need some sample aliases that follow R naming conventions
  mutate(
    alias = case_when(
      genotype == "WT" ~ paste0("WT_", sample),
      genotype == "EOfAD-like" ~ paste0("fAD_", sample),
      genotype == "MPS-IIIB" ~ paste0("MPS_", sample),
    ),
    genotype = factor(genotype, levels = c("WT", "EOfAD-like", "MPS-IIIB")),
    group = genotype
  ) %>%
  dplyr::arrange(genotype) %>%
  dplyr::select(sample, genotype, alias, everything())
```

```{r genoCols}
genoCols <- metadata$genotype %>%
  levels() %>%
  length() %>%
  brewer.pal("Set1") %>%
  setNames(levels(metadata$genotype))
```

```{r samples_byGeno}
samples_byGeno <- metadata %>%
  split(f = .$genotype) %>%
  sapply(function(x){
    pull(x, sample)
  }, simplify = FALSE)
```

## Differential expression data

```{r dgeList}
dgeList <- readRDS(here("files/dgeList.Rds"))
```

```{r design}
design <- model.matrix(~0 + group, data = dgeList$samples) %>%
  set_colnames(c("WT", "EOfAD", "MPSIIIB"))
```

```{r contrasts}
contrasts <- makeContrasts(
  EOfADvsWT = EOfAD - WT,
  MPSIIIBvsWT = MPSIIIB - WT,
  levels = colnames(design)
)
```

```{r topTables}
topTables <- readRDS(here("files/topTables.Rds")) %>%
  set_names(c("EOfADvsWT", "MPSIIIBvsWT"))
```

```{r deGenes}
deGenes <- lapply(topTables, function(tt){
  dplyr::filter(tt, DE) %>%
    pull(gene_id)
})
```

## geneDAR

```{r}
geneDAR <- readRDS(here("files/masterGR.Rds")) %>%
  map2(topTables, function(x, y){
    left_join(as_tibble(x), y[c("gene_id", "logFC", "PValue", "DE")]) %>%
      GRanges()
  }) %>%
  endoapply(function(x){plyranges::mutate(x, dar = diversity)}) %>%
  set_names(c("EOfADvsWT", "MPSIIIBvsWT"))
```

# topGO

```{r}
library(topGO)
library(org.Dr.eg.db)
library(GO.db)
```

```{r notes, eval=FALSE, include=FALSE}
## -- Notes to self for *.db objects --
## Show columns/keytypes
columns(org.Dr.eg.db)
## Get information on columns/keytypes
help("ENSEMBL")
```

```{r ens2GO}
ens2GO <- AnnotationDbi::select(
  org.Dr.eg.db,
  keys = keys(org.Dr.eg.db, "ENSEMBL"),
  columns = c("GO"),
  keytype = "ENSEMBL"
) %>%
  distinct(ENSEMBL, GO) %>%
  split(f = .$ENSEMBL) %>%
  lapply(extract2, "GO")
```

```{r GO_info}
GO_terms <- AnnotationDbi::select(
  GO.db,
  keys = keys(GO.db, "GOID"),
  columns = c("TERM", "DEFINITION"),
  keytype = "GOID"
) %>%
  as_tibble()
```

```{r}
all_genes <- topTables$MPSIIIBvsWT %>%
  with(structure(FDR, names = gene_id))
```

```{r}
sigGenes <- function(allScore) {
  return(allScore < 0.05)
}
```

```{r}
GO_data <- new(
  "topGOdata",
  ontology = "BP",
  allGenes = all_genes,
  geneSel = sigGenes,
  annot = annFUN.gene2GO,
  gene2GO = ens2GO,
  nodeSize = 5
)
```

```{r}
fisher <- runTest(GO_data, "classic", "fisher")
ks <- runTest(GO_data, "classic", "ks")
GenTable(GO_data, fisher = fisher, ks = ks, orderBy = "fisher") %>%
  left_join(GO_terms, by = c("GO.ID" = "GOID")) %>%
  as_tibble()
```

```{r}
all_genes <- topTables$MPSIIIBvsWT %>%
    # mutate(stat = -log10(PValue) * sign(logFC)) %>%
    mutate(stat = -log10(PValue)) %>%
    dplyr::arrange(desc(stat)) %>%
    with(structure(stat, names = gene_id))
# suppressMessages({
  tmp <- runTest(GO_data, "weight01", "ks")
# })
GenTable(GO_data, p = tmp, topNodes = 30) %>%
  dplyr::mutate(p = as.numeric(p), sig = p < 0.05)
```

